lib/internal/util.js

```diff
@@ -187,3 +191,65 @@ exports.convertToValidSignal = function convertToValidSignal(signal) {
 
   throw new Error('Unknown signal: ' + signal);
 };
+
+const kCustomPromisifiedSymbol = Symbol('util.promisify.custom');
+const kCustomPromisifyArgsSymbol = Symbol('customPromisifyArgs');
+
+function promisify(orig) {
+  if (typeof orig !== 'function') {
+    throw new TypeError('The first argument to promisify() must be a function');
+  }
+
+  if (orig[kCustomPromisifiedSymbol]) {
+    const fn = orig[kCustomPromisifiedSymbol];
```

Comment:
> This seems like it's observably installing something onto the original function

Onto an user-supplied promisified version of the original function, yes. I know this seems a bit odd, but it makes sense for consistency with the common case of not having a custom function…

> symbols are fully public

I know that :)

> If so, could this use a WeakMap as a cache instead?

It could, but are you worried about a performance impact, or visibility of the symbol? If it’s the former, I’m not sure that’s relevant, it would be odd for users to call their own function that they made specifically for `util.promisify()`; if it’s the latter… it’s a symbol meant for public use, not to hide something.